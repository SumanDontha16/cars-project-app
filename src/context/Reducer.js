/* eslint import/no-anonymous-default-export: [2, {"allowArrowFunction": true}] */
export default (state, action) => {
	switch (action.type) {
		case "CREATE_NEW_CAR":
			return {
				...state,
				car: [...state.car, action.payload],
			};

		case "UPDATE_CAR":
			const editingCar = action.payload;
			const updatedCar = state.car.map((car) => {
				if (car.id === editingCar.id) {
					return editingCar;
				}
				return car;
			});
			return {
				...state,
				car: updatedCar,
			};

		case "DELETE_CAR":
			return {
				...state,
				car: state.car.filter((car) => car.id !== action.payload),
			};

		default:
			return state;
	}
};
