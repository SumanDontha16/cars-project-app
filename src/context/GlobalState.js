import React, { createContext, useReducer } from "react";
import Reducer from "./Reducer";

const initialState = {
	car: [
		{
			id: 1,
			make: "Ford",
			model: "Puma",
			colour: "Black",
			year: "2009",
		},
	],
};

export const GlobalContext = createContext(initialState);
export const GlobalProvider = ({ children }) => {
	const [state, dispatch] = useReducer(Reducer, initialState);

	const createNewCar = (car) => {
		dispatch({
			type: "CREATE_NEW_CAR",
			payload: car,
		});
	};

	const deleteCar = (id) => {
		dispatch({
			type: "DELETE_CAR",
			payload: id,
		});
	};

	const updateCar = (car) => {
		dispatch({
			type: "UPDATE_CAR",
			payload: car,
		});
	};

	return (
		<GlobalContext.Provider
			value={{
				car: state.car,
				createNewCar,
				deleteCar,
				updateCar,
			}}
		>
			{children}
		</GlobalContext.Provider>
	);
};
