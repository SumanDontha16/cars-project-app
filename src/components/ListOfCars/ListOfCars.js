import React, { useContext } from "react";
import { Link } from "react-router-dom";
import { GlobalContext } from "../../context/GlobalState";

const ListOfCars = () => {
	const { car, deleteCar, updateCar } = useContext(GlobalContext);
	return (
		<>
			<div className="container">
				<div className="row">
					<div className="col-md-8"></div>
					<div className="col-md-4 text-right">
						{car.length >= 1 ? (
							<Link to="/create">
								<button className="btn btn-primary" type="button">
									Create New Car
								</button>
							</Link>
						) : (
							<p></p>
						)}
					</div>
				</div>
				<div>
					{car?.length > 0 ? (
						<div className="table-responsive">
							<table className="table table-striped table-hover">
								<thead>
									<tr>
										<th scope="col">Make</th>
										<th scope="col">Model</th>
										<th scope="col">Colour</th>
										<th scope="col">Year</th>
									</tr>
								</thead>
								<tbody>
									{car.map((c) => (
										<tr key={c.id}>
											<td>
												<Link to={`/detail/${c.id}`}>{c.model}</Link>
											</td>
											<td>
												<Link to={`/detail/${c.id}`}>{c.make}</Link>
											</td>
											<td>{c.colour}</td>
											<td>{c.year}</td>
											<td>
												<Link to={`/edit/${c.id}`}>
													<i
														className="bi bi-pencil-square"
														onClick={() => updateCar(c.id)}
													></i>
												</Link>
											</td>
											<td>
												<i
													className="bi bi-x-circle-fill"
													onClick={() => deleteCar(c.id)}
												></i>
											</td>
										</tr>
									))}
								</tbody>
							</table>
						</div>
					) : (
						<div className="noList">
							<p className="lead"> It seems we dont have any cars </p>
							<Link to="/create">
								<button type="button" className="btn btn-primary">
									Create New Car
								</button>
							</Link>
						</div>
					)}
				</div>
			</div>
		</>
	);
};

export default ListOfCars;
