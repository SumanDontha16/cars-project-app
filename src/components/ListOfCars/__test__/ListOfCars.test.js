import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ListOfCars from "../ListOfCars";
import { MemoryRouter } from "react-router-dom";
import { GlobalProvider } from "../../../context/GlobalState";

describe("List Of Cars", () => {
	test("Component rendered successfully!", () => {
		const { container } = render(
			<GlobalProvider>
				<MemoryRouter>
					<ListOfCars />
				</MemoryRouter>
			</GlobalProvider>
		);
		expect(container).toMatchSnapshot();
	});

	test("Should show create new car button to add list", () => {
		render(
			<GlobalProvider>
				<MemoryRouter>
					<ListOfCars />
				</MemoryRouter>
			</GlobalProvider>
		);

		const button = screen.getByRole("button", { name: "Create New Car" });
		userEvent.click(button);
	});

	test("Table Headers rendered successfully!'", () => {
		render(
			<GlobalProvider>
				<MemoryRouter>
					<ListOfCars />
				</MemoryRouter>
			</GlobalProvider>
		);

		expect(screen.getByRole("table")).toBeInTheDocument();
		expect(
			screen.getByRole("columnheader", { name: "Make" })
		).toBeInTheDocument();
		expect(
			screen.getByRole("columnheader", { name: "Model" })
		).toBeInTheDocument();
		expect(
			screen.getByRole("columnheader", { name: "Colour" })
		).toBeInTheDocument();
		expect(
			screen.getByRole("columnheader", { name: "Year" })
		).toBeInTheDocument();
	});

	test("Table data rendered successfully!'", () => {
		render(
			<GlobalProvider>
				<MemoryRouter>
					<ListOfCars />
				</MemoryRouter>
			</GlobalProvider>
		);

		expect(screen.getByRole("cell", { name: "Puma" })).toBeInTheDocument();
		expect(screen.getByRole("cell", { name: "Ford" })).toBeInTheDocument();
		expect(screen.getByRole("cell", { name: "Black" })).toBeInTheDocument();
		expect(screen.getByRole("cell", { name: "2009" })).toBeInTheDocument();
	});

	test("Edit link working successfully!", () => {
		render(
			<GlobalProvider>
				<MemoryRouter>
					<ListOfCars />
				</MemoryRouter>
			</GlobalProvider>
		);
		userEvent.click(screen.getByRole("link", { name: "" }));
	});

	test("Detail link working successfully!", () => {
		render(
			<GlobalProvider>
				<MemoryRouter>
					<ListOfCars />
				</MemoryRouter>
			</GlobalProvider>
		);

		userEvent.click(screen.getByRole("link", { name: "Puma" }));
	});
});
