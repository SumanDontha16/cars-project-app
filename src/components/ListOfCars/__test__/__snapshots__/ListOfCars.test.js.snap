// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Of Cars Component rendered successfully! 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-8"
      />
      <div
        class="col-md-4 text-right"
      >
        <a
          href="/create"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Create New Car
          </button>
        </a>
      </div>
    </div>
    <div>
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-hover"
        >
          <thead>
            <tr>
              <th
                scope="col"
              >
                Make
              </th>
              <th
                scope="col"
              >
                Model
              </th>
              <th
                scope="col"
              >
                Colour
              </th>
              <th
                scope="col"
              >
                Year
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <a
                  href="/detail/1"
                >
                  Puma
                </a>
              </td>
              <td>
                <a
                  href="/detail/1"
                >
                  Ford
                </a>
              </td>
              <td>
                Black
              </td>
              <td>
                2009
              </td>
              <td>
                <a
                  href="/edit/1"
                >
                  <i
                    class="bi bi-pencil-square"
                  />
                </a>
              </td>
              <td>
                <i
                  class="bi bi-x-circle-fill"
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
